notebook/notebook/notebook.py

```diff
@@ -113,201 +143,334 @@ def start_pod(jupyter_token, image):
                 'app': 'notebook-worker',
                 'hail.is/notebook-instance': INSTANCE_ID,
                 'uuid': pod_id,
-            }),
+                'svc_name': svc.metadata.name,
+                **labels
+            },),
         spec=pod_spec)
     pod = k8s.create_namespaced_pod(
         'default',
         pod_template,
-        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS,
+    )
+
     return svc, pod
 
 
-def external_url_for(path):
-    # NOTE: nginx strips https and sets X-Forwarded-Proto: https, but
-    # it is not used by request.url or url_for, so rewrite the url and
-    # set _scheme='https' explicitly.
-    protocol = request.headers.get('X-Forwarded-Proto', None)
-    url = flask.url_for('root', _scheme=protocol, _external='true')
-    return url + path
+def del_pod(pod_name: str):
+    k8s.delete_namespaced_pod(
+        pod_name,
+        'default',
+        kube.client.V1DeleteOptions(),
+        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+
+
+def del_svc(svc_name: str):
+    k8s.delete_namespaced_service(
+        svc_name,
+        'default',
+        kube.client.V1DeleteOptions(),
+        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+
+
+# General resource marshalling functions #
+
+
+def get_path(data, path: str):
+    # Not using tail recursion because python doesn't optimize such calls
```

Comment:
I believe we should elide this comment. This is a basic part of python semantics that we should verify in or teach to new hires.

Reply:
Sure